### PR TITLE
Stacks chainstate refactor for PoX

### PR DESCRIPTION
**Pull request by jcnelson**
This PR implements #1677 and #1566.  It does so by systemically changing the way we index Stacks blocks, from using `burn_header_hash/block_hash` to `consensus_hash/block_hash`.  In doing so, Stacks blocks are unique to their PoX fork, and are linked not to the burnchain block that selected them, but to the history of accepted burnchain operations that lead to the sortition that selected it (represented by the consensus hash).

This PR does _not_ update the consensus hash construction to include the PoX anchor block bit-vector, and this PR does _not_ change any network protocols.  But, the PR does mark intended network protocol changes with `TODO(PoX)` comments.

EDIT: minor and hopefully obvious note -- because this PR changes the way we index blocks, the MARF root hash calculation is different.  This PR is _not_ be compatible with the current Argon testnet.

### CODE REVIEW

File: src/chainstate/burn/db/sortdb.rs [hunk 1]
```diff
@@ -668,6 +675,19 @@ impl <'a> SortitionHandleConn <'a> {
...
+    pub fn open_reader_consensus_stubbed(connection: &'a SortitionDBConn<'a>, chain_tip: &ConsensusHash) -> Result<SortitionHandleConn<'a>, db_error> {
+        let sn = match SortitionDB::get_block_snapshot_consensus(&connection.conn, chain_tip)? {
+            Some(sn) => sn,
+            None => {
+                test_debug!("No such chain tip consensus hash {}", chain_tip);
+                return Err(db_error::NotFoundError);
+            }
+        };
+
+        SortitionHandleConn::open_reader_stubbed(connection, &sn.burn_header_hash)
```
Comment (kantai) [inline src/chainstate/burn/db/sortdb.rs:689]: This shouldn't need to be marked stubbed -- a reader could be opened with `&sn.sortition_id`.

**Comment by jcnelson** [comment]
@kantai I had couple of questions for you about the role you see `SortitionId` playing, now that I've had a chance to do some deep refactoring work for PoX.

First, do we need `SortitionId` at all?  If all we need from it is a globally-unique identifier that distinguishes two sortitions in the same burnchain block by which PoX fork they occurred on, then wouldn't the `ConsensusHash` be sufficient?  The `ConsensusHash` will be derived from the history of PoX-selected burnchain operations that lead to the given sortition, including the history of which PoX anchor blocks that have (or have not) arrived on the node.

Second, is there a need a sortition-history-independent identifier (e.g. `SortitionId`) that is distinct from a sortition-history-dependent identifier (e.g. `ConsensusHash`)?  If not, then going forward, maybe we don't need a PoX fork table.  Instead, maybe we could represent the PoX fork state using a standard MARF, where the `MarfTrieId` for a sortition is derived from its consensus hash?  We'd still need to write the `ChainsCoordinator` logic such that only PoX histories that contain all known anchored blocks will be considered valid to build on, but I think that's doable.

Thoughts?

**Comment by kantai** [comment]
I think that you may be right -- we could identify sortitions using a consensus hash that periodically mixes in a PoX presence bit (we'd just pad the consensus hash into a MarfTrieId).

I think that with that, we could also probably get rid of the PoxDB -- what we'd need to do is store something along the lines of 2 key-value pairs at the start of every Pox reward cycle like:

```
sortition_db::pox_anchor_presence::<consensus-hash>::<stacks-block-hash> -> (was the anchor for this reward cycle present)?
sortition_db::pox_anchor_start_consequences::<consensus-hash>::<stacks-block-hash> -> (consensus hash of the sortition where this reward cycle starts)
```

This would let the ChainsCoordinator:

1. Check whether or not a newly processed Stacks block was a PoxAnchor.
2. Determine the sortition that it needs to reset it's cursor to and begin "reprocessing" the sortitions from.

File: src/chainstate/burn/db/sortdb.rs (hunk 1 above)
Comment (jcnelson) [reply src/chainstate/burn/db/sortdb.rs:687]: Fixed

**Comment by jcnelson** [comment]
* Add a table of (consensus hash, block hash) to track the canonical PoX fork, as a "look-aside buffer" to the MARF state
   * maintained by chains coordinator -- invalidated and re-filled on PoX reorg
   * start out as a 1-row variant that just maintains the tip (so we can do ancestor queries on the canonical PoX fork)
* Networking code should not accept a Stacks block unless it's on the canonical PoX fork
   * always accept if it's a missing anchor block
   * need a way to efficiently reject a block based on its consensus hash -- i.e. if it's on a non-canonical PoX fork
* when processing a staging block, need to query the sortition DB MARF to see if it's on the canonical PoX fork
* chains coordinator should move blocks back into staging if they are no longer on the canonical PoX fork
* staging block processing can discard staging blocks that are not on the valid PoX fork

**Comment by psq** [comment]
> Networking code should not accept a Stacks block unless it's on the canonical PoX fork

maybe I misunderstand what's called the `canonical PoX fork`, but wouldn't that be too drastic?  For example, if a miner created a new block, but 3 others mined a longer chain with 2 more blocks off an older node, that miner would reject all these blocks and would not be able to re-join that longer fork.

**Comment by jcnelson** [comment]
> maybe I misunderstand what's called the canonical PoX fork, but wouldn't that be too drastic? For example, if a miner created a new block, but 3 others mined a longer chain with 2 more blocks off an older node, that miner would reject all these blocks and would not be able to re-join that longer fork.

Excellent question.  The PoX best-fork criteria are different from other blockchain fork criteria in that the best PoX fork is the one with the most-complete prefix of PoX anchor blocks.

Suppose you have two nodes, with the following two sets of anchor blocks:

```
A: [0] --> [1] --> [2] --> [3]
B: [0] --> [x] --> [2'] --> [3'] --> [4'] --> [5']
```

Here, node B is missing anchor block 1 (denoted as `[x]`).  Despite B having more anchor blocks than A, B's PoX fork is worse than A's, since the fact that B doesn't have anchor block `[1]` means that B's anchor blocks `[2']`, `[3']`, `[4']`, and `[5']` may not, in fact, be anchor blocks (and may be different form A's anchor blocks `[2]` and `[3]`).  This is because anchor block `[1]` will have determined an alternative sortition history for the remainder of the chain that B doesn't know about.  This may mean that once B discovers and processes anchor block `[1]`, B's subsequent anchor blocks will not have won their respective sortitions.  Therefore, B's fork has less information than A's fork, and is thus worse.  (It's not guaranteed that B's anchor blocks will be invalidated once it discovers `[1]`, however -- for example, the same anchor blocks `[2]` through `[5]` may have been mined on two different PoX forks -- but it _does_ mean that B needs to discard all the chain history after `[x]` and re-process their sortitions and blocks once it discovers and processes `[1]`).

Similarly, A's anchor block is better than B in this case as well:

```
A: [0] --> [x] --> [2] --> [3] --> [4]
B: [0] --> [x] --> [2] --> [x] --> [4'] --> [5']
```

This is because both A and B agree on the presence (and absence) of the same anchor blocks up to anchor block `[2]`, but because B is missing anchor block `[3]`, the remainder of its PoX fork should not be considered valid by A.

What this means in the networking code is that before node A accepts a block pushed to it from node B, node A needs to make sure that node B agrees with it on the shortest PoX anchor block history that contains the pushed block.  Otherwise, node A should reject the pushed block, since node B doesn't even agree on the same sortition history as A (meaning that A won't recognize the block as corresponding to a known sortition).

**Comment by psq** [comment]
thanks @jcnelson this is somewhat close to what I was understanding, and now it is clearer.

The case I was thinking about, which seems to be happening all the time right now (separate issue coming soon), is more something like this:

The blocks look like this (`[b-s]` notation means burn block `s`, stacks block height `s`:
```
[0-0] --> [1-1] --> [2-2] --> [4-3]
          \-> [3-1] --> [5-2] --> [6-3] --> [7-4]
```
assuming a miner has only:
```
[0-0] --> [1-1] --> [2-2] --> [4-3]
          \-> [3-1] -->
```
its canonical fork is the one with the `[4-3]` tip, so when `[5-2`] comes in, per the statement above, as it is not on the canonical chain, it will be rejected, which would preclude the miner to ever get the actual canonical fork, which tip is `[7-4]` if I understood correctly.

**Comment by jcnelson** [comment]
@psq Per your example, `[5-2]` would accepted by the remote node as long as it builds off of a valid PoX anchor block fork.  There's three levels of forking going on -- there's the Bitcoin chain histories, and within each Bitcoin history there are one or more PoX anchor block histories, and within a PoX anchor block history there are one or more Stacks block histories.  You can think of the current testnet as the special case where the only valid PoX history is the one without any anchor blocks (i.e. no anchor blocks exist).  The PoX anchor block history is "in-between" its Bitcoin history and the Stacks histories -- Stacks fork histories on an invalid PoX anchor block history are invalid.

The use of the term "invalid" as opposed to "non-canonical" to describe PoX forks is intentional -- a PoX fork that is missing one or more anchor blocks that are known to exist _can never be valid_.  This is different from how both Bitcoin and Stacks forks work, where a non-canonical fork can overtake the canonical fork later.  Put another way, the arrival of a PoX anchor block invalidates all PoX forks that do not have it.  In doing so, this would invalidate all Stacks forks that were built on those now-invalid PoX forks.

PoX forks are scoped to a Bitcoin fork.  There is at most one valid PoX fork per Bitcoin fork, so the node will track multiple valid but non-canonical PoX forks as part of tracking Bitcoin forks.  The canonical valid PoX fork is the valid PoX fork on canonical Bitcoin fork.

So, the validation logic for a Stacks block arrival would be to:
* Determine which Stacks history or histories it attaches to (there can be more than one; see below (*)).  If it doesn't attach to any, then reject.
* If the Stacks block is a PoX anchor block, then accept it and invalidate any conflicting PoX histories and their contained Stacks histories.
* If the Stacks block is NOT a PoX anchor block, then determine if the Stacks history (or histories) it is part of is/are on the valid PoX history.  If not, then reject.

(*) If the Bitcoin chain itself forks, then a Stacks block can potentially be attached to the PoX histories on both forks.  But, this would only happen if the Stacks block is mined on the first children of the block from which both forks descend (i.e. the block-commit transaction is mined in both Bitcoin forks and it wins sortition in both cases).  All subsequent descendants would be necessarily unique due to the way the VRF proof is constructed (i.e. the proof seed contains the hash of the Bitcoin block header that selected the Stacks block's parent).

**Comment by jcnelson** [comment]
> Looks good to me, thank you @jcnelson!
> Quick suggestion concerning the PoX effort in general, should we update SIP007 as we go?
> For instance, a note unpacking the canonical fork definition update (initially defined in SIP001 https://github.com/blockstack/stacks-blockchain/blob/master/sip/sip-001-burn-election.md#fork-selection) could be useful.

Yes, we should do our best to keep SIP-007 in sync with what the code does.  SIP-001 should be updated with a forward link to SIP-007 to describe how fork selection occurs with PoX, so that the reader can understand that PoB is a special case.  Moreover, the fork selection rule in SIP-001 is correct _within the context of a valid PoX anchor block fork_, but per my comment above, the presence of PoX anchor blocks adds another forking layer in-between the Stacks chain and Bitcoin chain.

**Comment by psq** [comment]
thank you for the detailed explanation @jcnelson, what you initially said makes sense now.

At the time I asked the question above, I was not quite clear on the notion of `PoX anchor block`, and I was taking it to be the same as a Stacks anchor block, which happens a lot more often, and since then I've refined my understanding by re-reading the SIPs and some related issues.

Bit by bit, the whole picture is getting clearer...

**Pull request merged**